source/shared/core_conn.cpp

```diff
@@ -932,6 +937,70 @@ void determine_server_version( _Inout_ sqlsrv_conn* conn TSRMLS_DC )
     conn->server_version = version_major;
 }
 
+void load_azure_key_vault( _Inout_ sqlsrv_conn* conn TSRMLS_DC )
+{
+    // If column encryption is not enabled simply do nothing. Otherwise, check if Azure Key Vault
+    // is required for encryption or decryption. Note, in order to load and configure Azure Key Vault, 
+    // all fields in conn->ce_option must be defined. 
+    if ( ! conn->ce_option.enabled || ! conn->ce_option.akv_required )
+        return;
+    
+    CHECK_CUSTOM_ERROR( conn->ce_option.akv_auth == NULL || Z_STRLEN_P(conn->ce_option.akv_auth) <= 0, conn, SQLSRV_ERROR_AKV_AUTH_MISSING) {
+        throw core::CoreException();
+    }
+
+    CHECK_CUSTOM_ERROR( conn->ce_option.akv_id == NULL || Z_STRLEN_P(conn->ce_option.akv_id) <= 0, conn, SQLSRV_ERROR_AKV_NAME_MISSING) {
+        throw core::CoreException();
+    }
+
+    CHECK_CUSTOM_ERROR( conn->ce_option.akv_secret == NULL || Z_STRLEN_P(conn->ce_option.akv_secret) <= 0, conn, SQLSRV_ERROR_AKV_SECRET_MISSING) {
+        throw core::CoreException();
+    }
+    
+    char *akv_auth = Z_STRVAL_P( conn->ce_option.akv_auth );
+    char *akv_id = Z_STRVAL_P( conn->ce_option.akv_id );
+    char *akv_secret = Z_STRVAL_P( conn->ce_option.akv_secret );    
+    unsigned int id_len = static_cast<unsigned int>( Z_STRLEN_P( conn->ce_option.akv_id ));
+    unsigned int key_size = static_cast<unsigned int>( Z_STRLEN_P( conn->ce_option.akv_secret ));    
+      
+    if ( !stricmp(akv_auth, "KeyVaultPassword") )
+    {
+        configure_azure_key_vault( conn, AKV_CONFIG_FLAGS, AKVCFG_AUTHMODE_PASSWORD, 0 );
+    }
+    else if ( !stricmp(akv_auth, "KeyVaultClientSecret") )
+    {
+        configure_azure_key_vault( conn, AKV_CONFIG_FLAGS, AKVCFG_AUTHMODE_CLIENTKEY, 0 );
+    }
+    
+    configure_azure_key_vault( conn, AKV_CONFIG_PRINCIPALID, akv_id, id_len );
+    configure_azure_key_vault( conn, AKV_CONFIG_AUTHSECRET, akv_secret, key_size );
+}
+
+void configure_azure_key_vault( sqlsrv_conn* conn, BYTE config_attr, const DWORD config_value, size_t key_size )
+{
+    BYTE akv_data[sizeof( CEKEYSTOREDATA ) + sizeof(DWORD) + 1 ];
+    CEKEYSTOREDATA *pData = reinterpret_cast<CEKEYSTOREDATA*>( akv_data );
+    pData->name = L"AZURE_KEY_VAULT";
+    pData->data[0] = config_attr;
+    pData->dataSize = sizeof(config_attr) + sizeof(config_value);
+    *reinterpret_cast<DWORD*>(&pData->data[1]) = config_value;
+    
+    core::SQLSetConnectAttr( conn, SQL_COPT_SS_CEKEYSTOREDATA, reinterpret_cast<SQLPOINTER>(pData), SQL_IS_POINTER );
+}
+
+void configure_azure_key_vault( sqlsrv_conn* conn, BYTE config_attr, const char* config_value, size_t key_size )
+{
+    BYTE akv_data[sizeof( CEKEYSTOREDATA ) + 2048 ];
+    CEKEYSTOREDATA *pData = reinterpret_cast<CEKEYSTOREDATA*>( akv_data );
+    pData->name = L"AZURE_KEY_VAULT";
+    pData->data[0] = config_attr;
+    pData->dataSize = 1+key_size;
+
```

Comment:
I don't think you actually need to add 1 to dataSize so long as the string is null-terminated. Please double check.

Reply:
I think it is necessary, it's done in the original source and Z_STRLEN does not include the null terminator.